### ISSUE TITLE
Redundancy in BMW?

### ISSUE THREAD

**Issue by JMMackenzie** [pisa-engine/pisa#281]
A very minor issue in BMW where we have an `if` before a `std::min` which doesn't seem to be required? We can either do an `if` and then assign the smaller one, or we can just do a `std::min`, I don't see why we need both.

https://github.com/pisa-engine/pisa/blob/a046325f1b896142eaca469ecbf0d9c7bfa8d417/include/pisa/query/algorithm/block_max_wand_query.hpp#L143-L146

**Comment by elshize** [comment]
Yeah, I think you're right, seems unnecessary. But I'm not sure which one is better or whether it makes any difference in performance. Any opinions?

**Comment by JMMackenzie** [comment]
I'd almost bet they would compile down to the same instructions, so probably just using `std::min` is clearer? Let's see what @amallia thinks.

**Comment by elshize** [comment]
Yeah, it would be surprising to me if there was a difference after optimization.

**Comment by JMMackenzie** [comment]
I think we have another redundancy.

https://github.com/pisa-engine/pisa/blob/a046325f1b896142eaca469ecbf0d9c7bfa8d417/include/pisa/query/algorithm/block_max_wand_query.hpp#L136-L160

1. We set `next_jump` to be the maximum document identifier. [Line 137]
2. If we have at least one list beyond the pivot list, we take the docid in that list to be our `next_jump`. [Line 139-141]
3. We loop over the lists 0 -> pivot (inclusive) and pick up the smallest identifier bounding the current block configuration. [Line 143-146]
4. We set `next` to `next_jump + 1` to ensure we are exiting the current block configuration. [Line 148]
5. We go back now and check to make sure that we did not step over the docid from **step 2** above [Line 149-153]
6. Finally, we check that our new id is actually greater than the one we are skipping. If not, we ensure this is the case [Line 155-157]

I think, to summarize, we are doing things in a strange way. In step 2 above, we are making sure we check the list beyond the pivot list to pick up that identifier, and then we check again in step 5. Seems redundant. 

My proposal would be something like this, which removes the redundant check (step 2) and removes the requirement for two variables `next_jump` and `next`:

```
 uint64_t next = max_docid; 
  
 for (size_t i = 0; i <= pivot; ++i) { 
     if (ordered_cursors[i]->w.docid() < next_jump) 
         next = std::min(ordered_cursors[i]->w.docid(), next); 
 } 
  
 next = next + 1; 
 if (pivot + 1 < ordered_cursors.size()) { 
     if (next > ordered_cursors[pivot + 1]->docs_enum.docid()) { 
         next = ordered_cursors[pivot + 1]->docs_enum.docid(); 
     } 
 } 
  
 if (next <= ordered_cursors[pivot]->docs_enum.docid()) { 
     next = ordered_cursors[pivot]->docs_enum.docid() + 1; 
 } 
  
 ordered_cursors[next_list]->docs_enum.next_geq(next); 
```

We could also substitute calls to `ordered_cursors[pivot]->docs_enum.docid()` to `pivot_id` as it's set previously in the flow control.

If you both agree, I will prepare a PR to include these two improvements.

**Comment by JMMackenzie** [comment]
PS, regarding the `std::min` vs `if()` for finding mins, I ran a small benchmark that generated random integers and then iterated over them searching for the min. It seems that the `if` version is quite a lot faster after optimization (over huge vectors). Probably not enough to matter for our code/purposes, but it is what it is,

**Comment by elshize** [comment]
Strange about the benchmark. Just out of curiosity, what compiler and
standard library implementation are you using?

I mean, it is really not the same thing, because of that assignment. It
probably results in assembly that always assigns and that write is the
difference. Maybe. Regardless, thanks for benchmarking, we should just use
the if statement then.

**Comment by JMMackenzie** [comment]
GCC 9.2.1, it was just a quick 'n' dirty benchmark on my laptop. If you like, I can re-do it a bit better and report the results here on a few different machines?

**Comment by elshize** [comment]
Hold of on that, I want to check something first before you spend time on
it. I'll get back to you later tonight.

**Comment by amallia** [comment]
Can we measure of the effect of the proposed change?

**Comment by JMMackenzie** [comment]
Yes of course. I'll put together a test tomorrow with the two versions so we can compare, and I will report back here.

**Comment by elshize** [comment]
I checked the assembly of two functions: https://gcc.godbolt.org/z/2k90uO on both GCC and clang, and when using `min` function, both generate `cmp`, then one conditional move between registers followed by an unconditional move to from register to memory. I imagine that conditional move is fast, and that's the optimized version of `min` itself. In the `if` case, you got your standard `cmp`, followed by conditional jump to `ret` and the assignment in between. Clang additionally compares directly with memory rather than to copy value to a register first (only `if` version) but I have no idea if that makes any difference or not.

This one is tough for me to judge. On one hand, you got that unconditional move to memory, on the other the conditional jump, but maybe since the jump wastes at most one cycle (is that even true? I don't know), and then some of those times will predict the right branch anyway, then maybe that's the more efficient way.

Bottom line is that one need to test to make sure, but also I don't think the difference should be significant anyway...

That said, by time I finished writing, it occurred to me that I haven't compiled the original version with `if` and `min`, and sure enough, both gcc and clang are smart enough to optimize away `min` and produce the exact same output as `if` alone ;) https://gcc.godbolt.org/z/oKfJlq

**Comment by JMMackenzie** [comment]
Good to know, thanks for checking. 

Nevertheless, I'll compare the other issue I raised and see if we can simplify the code a little. I can also try the two different min methods and see if anything turns up at all. That branch probably isn't taken often enough to make a real difference anyway, but we can see.

**Comment by elshize** [comment]
By the way, and this shouldn't affect performance here but just wanted to mention, if we have an `if` statement where we execute a function twice, once in the condition, and then later in the body, we can use the c++17 syntax:
```cpp
if (auto first_beyond_pivot = ordered_cursors[pivot + 1]->docs_enum.docid(); next > first_beyond_pivot) { 
    next = first_beyond_pivot; 
}
```
If the access is more complicated then it can clearly speeds things up, but even if it doesn't we sometimes might want to do it for the sake of naming the variable and therefore being more explicit about what we're doing.

Although none of these two really apply to this situation, I argue that even here it is better, because repeating the same value is error prone, what if you forget `+ 1`, then even if someone catches it, it will not be apparent right away that this is not intended. But once you extract that value once, and then just compare/assign names, it's clear and you can't screw it up below.

**Comment by JMMackenzie** [comment]
On a relatively large log of varying query lengths on Gov2 with SIMDBP:

Master: `{"type": "block_simdbp", "query": "block_max_wand", "avg": 22576.9, "q50": 15014, "q90": 49336, "q95": 69548}`

Suggested: `{"type": "block_simdbp", "query": "block_max_wand", "avg": 22530.1, "q50": 14990, "q90": 49264, "q95": 69539}`

I'd say differences are just by chance.

I am comparing these two blocks in particular:

Master:
```
                uint64_t next_jump = max_docid;

                if (pivot + 1 < ordered_cursors.size()) {
                    next_jump = ordered_cursors[pivot + 1]->docs_enum.docid();
                }

                for (size_t i = 0; i <= pivot; ++i) {
                    if (ordered_cursors[i]->w.docid() < next_jump)
                        next_jump = std::min(ordered_cursors[i]->w.docid(), next_jump);
                }

                next = next_jump + 1;
                if (pivot + 1 < ordered_cursors.size()) {
                    if (next > ordered_cursors[pivot + 1]->docs_enum.docid()) {
                        next = ordered_cursors[pivot + 1]->docs_enum.docid();
                    }
                }

                if (next <= ordered_cursors[pivot]->docs_enum.docid()) {
                    next = ordered_cursors[pivot]->docs_enum.docid() + 1;
                }
```
Suggested:
```
                next = max_docid;

                for (size_t i = 0; i <= pivot; ++i) {
                    if (ordered_cursors[i]->w.docid() < next)
                        next = ordered_cursors[i]->w.docid();
                }

                next = next + 1;
                if (pivot + 1 < ordered_cursors.size() && ordered_cursors[pivot + 1]->docs_enum.docid() < next) {
                    next = ordered_cursors[pivot + 1]->docs_enum.docid();
                }

                if (next <= pivot_id) {
                    next = pivot_id + 1;
                }
```

**Comment by elshize** [comment]
In that case, I'd go with the simplified version.
